### PR TITLE
MAINT: check `_append_char` return value

### PR DESCRIPTION
**Pull request by mattip**
Add missing check for return value of _append_char due to `malloc` failure

### CODE REVIEW

**Review by jeffyancey — APPROVED** [review]
These changes look good to me. @eric-wieser, would you mind expanding briefly on your comment, it wasn't immediately apparent to me what you meant.

**Comment by eric-wieser** [comment]
If a function returns `-1` on error, we almost always write `if (some_func(...) < 0) { /* fail */ }`.

File: numpy/core/src/multiarray/buffer.c [hunk 1]
```diff
@@ -134,7 +134,7 @@ static int
 _append_str(_tmp_string_t *s, char const *p)
 {
     for (; *p != '\0'; p++) {
-        if (_append_char(s, *p) != 0) {
+        if (_append_char(s, *p)) {
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/core/src/multiarray/buffer.c]: Refactored accordingly

File: numpy/core/src/multiarray/buffer.c [hunk 2]
```diff
@@ -215,19 +215,19 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
...
 
             PyOS_snprintf(buf, sizeof(buf), "%ld", (long)dim_size);
-            _append_str(str, buf);
+            if (_append_str(str, buf) < 0) return -1;
             total_count *= dim_size;
         }
-        _append_char(str, ')');
+        if (_append_char(str, ')') < 0) return -1;
 
         Py_DECREF(subarray_tuple);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/buffer.c]: Memory leak here if the new error path is taken.

Comment (charris) [reply numpy/core/src/multiarray/buffer.c]: And we never put the statement on the same line :)
```
        if (_append_char(str, ')') < 0) {
            return -1;
        }
```

**Review by eric-wieser — APPROVED** [review]
LGTM. I think breaking into 3 lines would hinder clarity in most places.

**Comment by mattip** [comment]
@charris - style decision needed whether to revert 27834b3

**Comment by eric-wieser** [comment]
Let's just put this in - if we really care about style, we should starting running automated tools to enforce it.

**Comment by vanossj** [comment]
@mattip beat me to checking `_append_char`, but I think `tmp` could still leak, see #12035

**Pull request merged**